### PR TITLE
fix: introduce approx size to patterns and use in ScatterBrush

### LINKED ISSUE #2610: LocalVectorSet can only contain vectors within 1024 blocks (cuboid) of the first entry error.
### Server Implementation

Paper

### Server Version

1.20.4

### Describe the bug

When trying to use `/brush scatter #fullcopy[<schematic goes here>]` this error occurs

```
>....
[10:03:13 WARN]: java.lang.UnsupportedOperationException: LocalVectorSet can only contain vectors within 1024 blocks (cuboid) of the first entry. Attempted to set block at 10044, 640, 10251. With origin 10044 128 10251
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.fastasyncworldedit.core.math.LocalBlockVectorSet.add(LocalBlockVectorSet.java:303)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.fastasyncworldedit.core.command.tool.brush.ScatterBrush.build(ScatterBrush.java:84)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.sk89q.worldedit.command.tool.BrushTool.act(BrushTool.java:481)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.sk89q.worldedit.command.tool.BrushTool.actPrimary(BrushTool.java:367)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.sk89q.worldedit.extension.platform.PlatformManager.lambda$handlePlayerInput$6(PlatformManager.java:501)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.sk89q.worldedit.extension.platform.AbstractPlayerActor.lambda$runAction$1(AbstractPlayerActor.java:672)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$run$0(AsyncNotifyKeyedQueue.java:48)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$call$1(AsyncNotifyKeyedQueue.java:58)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
>....
[10:03:13 WARN]: 	at FastAsyncWorldEdit-Bukkit-2.9.0.jar//com.fastasyncworldedit.core.util.task.KeyQueuedExecutorService$KeyRunner.lambda$run$0(KeyQueuedExecutorService.java:150)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
>....
[10:03:13 WARN]: 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

### To Reproduce

1. Create a 1024 height world
2. Use /brush scatter #fullcopy[<schematic goes here>]` at above Y=384

### Expected behaviour

Schematics should just be placed.

### Screenshots / Videos

_No response_

### Error log (if applicable)

_No response_

### Fawe Debugpaste

https://athion.net/ISPaster/paste/view/92eae493c8c14635b3e5c43cf099b640

### Fawe Version

2.9.1-SNAPSHOT-670;9fa3385

### Checklist

- [X] I have included a Fawe debugpaste.
- [X] I am using the newest build from https://ci.athion.net/job/FastAsyncWorldEdit/ and the issue still persists.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dordsor21**
- fixes #2610

### CODE REVIEW

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/command/tool/brush/ScatterBrush.java [hunk 1]
```diff
@@ -64,8 +65,9 @@ public void build(EditSession editSession, BlockVector3 position, Pattern patter
             length = 1;
             visited.add(position);
         }
-        LocalBlockVectorSet placed = new LocalBlockVectorSet();
-        placed.setOffset(position.getX(), position.getZ());
+        BlockVector3 patternSize = pattern.size();
+        BlockVector3Set placed = BlockVector3Set.getAppropriateVectorSet(patternSize.add(distance, distance, distance));
+        placed.setOffset(position.getX(), (position.getY() + patternSize.getY()) / 2, position.getZ());
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/fastasyncworldedit/core/command/tool/brush/ScatterBrush.java:70]: That's probably meant to be 
```suggestion
        placed.setOffset(position.getX(), position.getY() + patternSize.getY() / 2, position.getZ());
```
but probably the pattern size isn't really beneficial here at all as it's not clear if it's all above or all below or something in between?

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/math/BlockVectorSet.java [hunk 2]
```diff
@@ -78,6 +78,49 @@ public boolean contains(int x, int y, int z) {
...
+        if (!contains(x + radius, y, z - radius)) {
+            return false;
+        }
+        // Slow but if someone wants to think of an elegant way then feel free to add it
+        for (int xx = -radius; xx <= radius; xx++) {
+            int rx = x + xx;
+            for (int yy = -radius; yy <= radius; yy++) {
+                int ry = y + yy;
+                for (int zz = -radius; zz <= radius; zz++) {
+                    if (contains(rx, ry, z + zz)) {
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/fastasyncworldedit/core/math/BlockVectorSet.java:115]: So if any block intersects, it's considered to contain the radius? Could we call the method `intersectsWithRadius` or something like that then instead?

**Review by SirYwell — CHANGES_REQUESTED** [review]

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/command/tool/brush/ScatterBrush.java (hunk 1 above)
Comment (dordsor21) [reply worldedit-core/src/main/java/com/fastasyncworldedit/core/command/tool/brush/ScatterBrush.java:70]: Yes that's true

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/math/BlockVectorSet.java (hunk 2 above)
Comment (dordsor21) [reply worldedit-core/src/main/java/com/fastasyncworldedit/core/math/BlockVectorSet.java:115]: I'm keeping the method name the same as the one that already existed in LocalBLockVectorSet

**Review by SirYwell — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**